## Problem Statement 

1. Build a model to predict sales if x amount of money is invested in TV, radio & newspaper advertising

2. Predict which mode of advertising is going to be effective?

3. If possible derive relations

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from pandas_profiling import ProfileReport

In [2]:
pip install pandas_profiling

Note: you may need to restart the kernel to use updated packages.


In [3]:
df = pd.read_csv('Advertising[1].csv')
df

Unnamed: 0     TV  Radio  Newspaper  Sales
0             1  230.1   37.8       69.2   22.1
1             2   44.5   39.3       45.1   10.4
2             3   17.2   45.9       69.3    9.3
3             4  151.5   41.3       58.5   18.5
4             5  180.8   10.8       58.4   12.9
..          ...    ...    ...        ...    ...
195         196   38.2    3.7       13.8    7.6
196         197   94.2    4.9        8.1    9.7
197         198  177.0    9.3        6.4   12.8
198         199  283.6   42.0       66.2   25.5
199         200  232.1    8.6        8.7   13.4

[200 rows x 5 columns]

In [4]:
df.describe()

Unnamed: 0          TV       Radio   Newspaper       Sales
count  200.000000  200.000000  200.000000  200.000000  200.000000
mean   100.500000  147.042500   23.264000   30.554000   14.022500
std     57.879185   85.854236   14.846809   21.778621    5.217457
min      1.000000    0.700000    0.000000    0.300000    1.600000
25%     50.750000   74.375000    9.975000   12.750000   10.375000
50%    100.500000  149.750000   22.900000   25.750000   12.900000
75%    150.250000  218.825000   36.525000   45.100000   17.400000
max    200.000000  296.400000   49.600000  114.000000   27.000000

In [5]:
pf = ProfileReport(df)

In [6]:
pf.to_widgets()

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render widgets:   0%|          | 0/1 [00:00<?, ?it/s]

Check for multi-collinearity between independent variables  and correlation b/w dependent and independent values

In [7]:
pf.to_file('test.html')

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [8]:
# selecting independent variable
x = df[["TV"]]
x

TV
0    230.1
1     44.5
2     17.2
3    151.5
4    180.8
..     ...
195   38.2
196   94.2
197  177.0
198  283.6
199  232.1

[200 rows x 1 columns]

In [9]:
# selecting dependent variable
y = df[["Sales"]]
y

Sales
0     22.1
1     10.4
2      9.3
3     18.5
4     12.9
..     ...
195    7.6
196    9.7
197   12.8
198   25.5
199   13.4

[200 rows x 1 columns]

In [10]:
from sklearn.linear_model import LinearRegression

In [11]:
linear = LinearRegression()

In [12]:
# fitting our training data into linear regression model
linear.fit(x,y)

LinearRegression()

In [13]:
linear.intercept_

array([7.03259355])

In [14]:
linear.coef_

array([[0.04753664]])

In [15]:
# ## Sales = m* TV + c || Sales = 0.0475x + 7.03

# file = 'linear_reg.sav'
# pickle.dump(linear, open(file, 'wb'))

In [16]:
linear.predict([[45]])

array([[9.17174237]])

In [17]:
## lets say they are following amount invested in TV ads 
# then we want to want what would be sales

l = [50,60,70,80,90,100]

for i in l:
    print(linear.predict([[i]]))


[[9.40942557]]
[[9.88479198]]
[[10.36015838]]
[[10.83552478]]
[[11.31089119]]
[[11.78625759]]


In [18]:
r_sq = linear.score(x,y) # it will give R^2i.e. accuracy of model
r_sq

0.611875050850071

In [19]:
## finding out Adjusted R square

num = (1- r_sq) * (len(df)-1)
den = len(df) - x.shape[1] - 1

In [20]:
adj_rsquare = 1 - (num/den)
adj_rsquare

0.6099148238341623

In [21]:
## Building multiple linear regression model

data_x = df[['TV', 'Radio', 'Newspaper']]
data_x

TV  Radio  Newspaper
0    230.1   37.8       69.2
1     44.5   39.3       45.1
2     17.2   45.9       69.3
3    151.5   41.3       58.5
4    180.8   10.8       58.4
..     ...    ...        ...
195   38.2    3.7       13.8
196   94.2    4.9        8.1
197  177.0    9.3        6.4
198  283.6   42.0       66.2
199  232.1    8.6        8.7

[200 rows x 3 columns]

In [22]:
y = df.Sales
y

0      22.1
1      10.4
2       9.3
3      18.5
4      12.9
       ... 
195     7.6
196     9.7
197    12.8
198    25.5
199    13.4
Name: Sales, Length: 200, dtype: float64

In [23]:
ln = LinearRegression()
ln.fit(data_x, y)

LinearRegression()

In [24]:
ln.intercept_

2.9388893694594085

In [25]:
ln.coef_

array([ 0.04576465,  0.18853002, -0.00103749])

In [26]:
# equation will be -> y = 0.045 * TV + 0.188 * radio - 0.0103 * newspaper

r_sq2 = ln.score(data_x,y) # accuracy
r_sq2

0.8972106381789522

In [27]:
## Adjusted R square

num = (1- r_sq2) * (len(df)-1)
den = len(df) - data_x.shape[1] - 1

In [28]:
adj_rsquare2 = 1 - num/den
adj_rsquare2

0.8956373316204668

1. Since accuracy is more more multiple regression hence we sjould consider this model

2. Adj R square has increased

In [29]:
data_x1 = df[['TV', 'Radio']]
data_x1

TV  Radio
0    230.1   37.8
1     44.5   39.3
2     17.2   45.9
3    151.5   41.3
4    180.8   10.8
..     ...    ...
195   38.2    3.7
196   94.2    4.9
197  177.0    9.3
198  283.6   42.0
199  232.1    8.6

[200 rows x 2 columns]

In [30]:
y = df[['Sales']]
y

Sales
0     22.1
1     10.4
2      9.3
3     18.5
4     12.9
..     ...
195    7.6
196    9.7
197   12.8
198   25.5
199   13.4

[200 rows x 1 columns]

In [31]:
ln2 = LinearRegression()
ln2.fit(data_x1, y)

LinearRegression()

In [32]:
r_sq3 = ln2.score(data_x1,y) # accuracy
r_sq3

0.8971942610828956

In [33]:
## Adjusted R square

num = (1 - r_sq3) * (len(df)-1)
den = len(df) - data_x1.shape[1] - 1

In [34]:
adj_rsquare3 = 1 - num/den
adj_rsquare3

0.8961505479974428

adjusted r square for TV feature - 0.6099148238341623
adjusted r square for TV + radio feature - 0.8961505479974428
adjusted r square for TV + radio + newspaper feature - 0.8956373316204668

This helps in feature selection as we can see that with 2 feature ajusted r sq has increased but with 3 features it has decreased. Therefore TV + radio is the most appropriate feature to consider